Implementation of faster MCMC CSV parsing and Stan CSV utilities

amas0 opened the pull request:
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

This is in an initial PR that addresses parts of #785. This introduces some general utilities for parsing Stan CSV files and uses some of those changes within `CmdStanMCMC`. 

Some key ideas:

* In an attempt to generalize some of the CSV parsing across different inference methods, this change uses the fact that the CSV outputs are made up of an sequences of commented/uncommented lines that correspond to different sections. The parser logic used here is state-based where a new parsing rule is used each time a transition occurs between a commented and un-commented section.
For example in the standard MCMC output, the sections are `config`, `warmup`, `adaptation`, `samples`, and `timing`. Then one only needs to write some parsing code for each section.
* An optimized parser based on `polars` is implemented that takes in `list[bytes]` corresponding to the lines in the CSV outputs that correspond to draws and produces a `np.array` as a drop in replacement for existing parsing.
* With these utilities, I introduce a `StanCsvMCMC` dataclass which represents the parsed output of a Stan CSV file. It includes a dict representing the sampler config (which just uses the existing `scan_config` function), the warmup draws as a numpy array (if present), the step size and mass matrix (if present), the sampling draws as a numpy array, and the timing information as a dictionary. The idea being that it will be easier to write and read code with this structured representation than have to know how the CSV files are structured.

I use these changes to update the `CmdStanMCMC._assemble_draws` method, which, with the optimized parsing, runs noticeably faster for large models.

Putting this PR as a draft for now, as I want to get some feedback and am planning on writing unit tests to cover the new code, but all existing unit tests are passing on this.

My plan for next steps is to implement a corresponding `StanCsv*` for the various inference methods (I've already done most of this because the logic is very similar) and then update how each of the Stan fit objects pull info from the CSV files. This will make things more consistent across the inference methods and hopefully clean up some parts of the library.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Myself



By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

amas0 committed:
Add StanCsvMCMC dataclass

This class is intended to serve as an abstraction layer between
the Stan fit objects (like CmdStanMCMC) and the Stan CSV output.
It converts the Stan CSV into structured format that can be used
to straightforwardly extract relevant info.

The practically important contribution here is the implementation
of a faster samples parser using pandas/pyarrow intended to replace
the pure-Python implementation.

Some of the structure/utilties included in this commit are intended
to clean up logic elsewhere in the library by sharing functionality
and making the processing of Stan CSV files more consistent across
the varying inference methods.

amas0 committed:
Filter out empty mass matrix lines

amas0 committed:
Update _assemble_draws to use StanCsvMCMC object

amas0 committed:
Fix code incompatible with Python 3.8

amas0 committed:
Convert draws parsing to polars

amas0 committed:
Add docstrings

amas0 committed:
Add initial unit tests

WardBrian commented on cmdstanpy/utils/stancsv.py:
```diff
@@ -1,17 +1,216 @@
+
+
+@dataclass
+class StanCsvMCMC:
+    """Class containing the parsed output of a Stan CSV file sourced
+    from the `sample` inference method."""
```
I'm a bit hesitant to introduce a new class per output type. I like the general look of this code and design, but I wonder if we couldn't flip it inside-out a bit to promote more re-use.

I'll try to elaborate on this in more comments below, but in short I wonder if the rules etc should be entirely in the `stanfit/` classes, and the code here much more generic

amas0 replied:
How about per output functions that handle the different comment lines present? Those can be defined in `utils/stancsv.py` and then within the `stanfit/` classes they just use those to handle extract the relevant In combination with the below comment. So within the `stanfit/mcmc.py` it could look like:

```python
with open(stan_csv, "rb") as f:
  comment_lines, draws: List[bytes], np.ndarray = stancsv.parse_general_stan_csv_from_lines(f)

config, step_size, mass_matrix, timings = stancsv.parse_mcmc_comment_lines(comment_lines)
```

I generally like the idea of the `stanfit/` classes not needing to concern themselves too much with IO logic, but some stuff like parsing of the adaptation comment blocks is unique to a particular inference method. At the very least logic for parsing the draws out, the config lines, and the timing lines all make sense to me to exist in the `utils/stancsv.py` space (as a side note, I think the config parsing that exists now probably should be cleaned up, but that'll be another day's task).

If we want to have the method-specific parsing logic, i.e. parsing the adaptation lines, to live within the `stanfit/` classes and the shared sections to use the shared utilities, I think that's reasonable.

Let me know what you think.

WardBrian replied:
That looks pretty good to me. Alt. would be `mass_matrix = stancsv.parse_mcmc_metric(comments[3])`, but the `3` feels arbitrary there.

amas0 replied:
At that point I imagine its not even worth separating out the comments into sections and just have a `comment_lines: list[bytes]` that gets passed around to various parsing methods. Could imagine something like:

```py
config = stancsv.parse_config(comments)
# I assume these should go together because they come from the same block
step_size, mass_matrix = stancsv.parse_mcmc_adaptation(comments)
timings = stancsv.parse_timings(comments)
```

Not sure, something like this probably. I'll play around with some changes and come back with updates when I'm happy. Thanks for the feedback.

WardBrian commented on cmdstanpy/utils/stancsv.py:
```diff
@@ -1,17 +1,216 @@
+                ParsingRule(action=adaptation_lines.append),
+                ParsingRule(action=sampling_lines.append),
+                ParsingRule(action=timing_lines.append),
+            )
```
The ParsingRules idea is neat, but all of them ultimate end up being list appending. What if instead the function returned `tuple(list[list[bytes]], np.ndarray)` (or a dataclass containing those, I don't mind).

This could then be used by all the interfaces, and it leaves it up to, e.g., stanfit.mcmc to decide that the third element of that first list should contain the metric or whatever. Right now we have logic for the metric here, and then a separate `if metric is None`-style check in the consumer of this code, but it feels like they could be combined easier.

This removes the distinction between warmup and sampling draws, but that's artificial anyway, as the condition up in mcmc.py (`self._save_warmup and parsed_csv.warmup_draws is not None`) should always either be both true or both false, meaning there's no need to really differentiate them.

amas0 replied:
Yeah I kind of had a feeling this approach was a little too clever. When I started going down this path, I assumed that there would be more variation in the parsing across the different inference methods, but it's mostly the same.

Splitting the parsed outputs into non-draws and draws would be simple and then any per-method parsing would happen against comment lines. I can rework this a bit.

WardBrian commented on cmdstanpy/utils/stancsv.py:
```diff
@@ -1,17 +1,216 @@
+        .astype(np.float32)
+    )
+    # Telling the type checker we know the type is correct
+    return cast(npt.NDArray[np.float32], out)
```
My instinct is still to make `polars` optional to match what we do with xarray and ujson

amas0 replied:
I went ahead and made `polars` optional with fallback logic to `np.loadtxt` if it isn't present.

WardBrian commented on cmdstanpy/utils/stancsv.py:
```diff
@@ -1,17 +1,216 @@
+    return cast(npt.NDArray[np.float32], out)
+
+
+def parse_hmc_adaptation_lines(
```
Just a comment to say I really like the style of these two functions here, much preferrable to the `scan_*` style we've had before

WardBrian reviewed (commented):
Thanks @amas0. The code looks really good. 

My comments below are thoughts on a slightly different style that I think will make it more reusable across the different inference algorithms and try to consolidate the logic for each in one place. Let me know what you think!

amas0 committed:
Make polars an optional dependency

amas0 committed:
Refactor parsing to be function-based

amas0 committed:
Add polars to test dependencies

amas0 committed:
Add single element csv parsing tests

amas0 committed:
Add fixed_param check before assembling draws

amas0 committed:
Add numpy/polars equiv testing

amas0 committed:
Convert tests from np.array_equiv to np.array_equal

amas0 committed:
Fix csv numpy parsing shape when single row

amas0 commented:
Okay, I think this PR should be good to go for more review/feedback. In response to the previous comments, I've made the following changes from the last version:

* Made `polars` an optional dependency with fallback-parsing done by `numpy`
* Added unit tests against the new `utils/stancsv` functions (and caught some bugs)
* Removed the `ParsingRule`s based parser in favor of a simpler structure that splits a CSV into comments/draws lines
* Comment section parsing functions like timing/adaptation parsing expect to be provided the full comment lines
* Remove `StanCsvMCMC` in favor of using the utility functions within the `CmdStanMCMC._assemble_draws()` method

This version is simpler and cleaner than the previous. I still think there's probably some cleanup that can happen in the `CmdStanMCMC._assemble_draws()` method, but at least I think it's improved compared to the existing version.

I ran one last bit of performance testing comparing the new `numpy`/`polars` based parsing against the existing method:

<img width="868" height="728" alt="image" src="https://github.com/user-attachments/assets/3442b55c-1be2-45e1-b673-0ea2a4f88473" />

This is with the defaults of 4 chains, 1000 draws per chain. I didn't stretch the test out to the extremes of number of parameters, but for moderate to large model sizes, the speedup should be noticeable.

WardBrian commented on cmdstanpy/utils/stancsv.py:
```diff
@@ -1,17 +1,150 @@
+
+def csv_bytes_list_to_numpy(
+    csv_bytes_list: List[bytes], includes_header: bool = True
+) -> npt.NDArray[np.float32]:
```
We should always load these as float64. With the default CSV output it's not that noticeable, but people can request all 16 digits if they wanted to.

amas0 replied:
Good catch -- updated. I had been working only off the default CSV and f32 seemed fine.

WardBrian commented on cmdstanpy/utils/stancsv.py:
```diff
@@ -1,17 +1,150 @@
+        try:
+            out = (
+                pl.read_csv(
+                    io.BytesIO(b"".join(csv_bytes_list)),
```
should I be surprised that this isn't `b'\n'`?

amas0 replied:
When using a file object as an iterator in Python, e.g. 

```py
with open('file.txt', 'r') as f:
  for line in f:
    foo(line)
```

the line terminators are kept at the end of each `line` (this is also true of `.readlines()`). So the partitioned lines all include `b"\n"` at the end already, so we just join on an empty byte to recreate the proper multi-line structure expected by the `read_csv`.

WardBrian commented on cmdstanpy/utils/stancsv.py:
```diff
@@ -1,17 +1,150 @@
+                ndmin=1,
+            )
+        if out.shape == (0,):
+            raise ValueError("No data found to parse") from exc
```
the `from exc` seems useless here. If it was caused by a pylint error you can safely `# ignore` it I think

amas0 replied:
Removed and added a pylint ignore comment

WardBrian commented on cmdstanpy/utils/stancsv.py:
```diff
@@ -1,17 +1,150 @@
+            out = out.reshape(1, -1)
+
+    # Telling the type checker we know the type is correct
+    return cast(npt.NDArray[np.float32], out)
```
Casts have a (small) runtime cost and generally defeat the purpose of typechecking. It's equivalent to `# type: ignore`, which I would prefer if we cant make the type work out some other way (e.g. by declaring it when `out` is defined?)

amas0 replied:
Should be fixed up now -- `cast` removed. Adding `npt.NDArray[np.float64]` to the declaration of `out` from the polars bit seems to have resolved it.

WardBrian commented on cmdstanpy/stanfit/mcmc.py:
```diff
@@ -429,83 +430,42 @@ def _assemble_draws(self) -> None:
-                    line = fd.readline().strip()
-                    xs = line.split(',')
-                    self._draws[i, chain, :] = [float(x) for x in xs]
+            with open(self.runset.csv_files[chain], "rb") as f:
+                comments, draws = stancsv.parse_stan_csv_comments_and_draws(f)
```
Should the `open` be inside the stancsv function?

amas0 replied:
This is something I went back and forth on. I could add in the relatively common interface where `parse_stan_csv_comments_and_draws` can accept `Union[str, Path, Iterator[bytes]]` (or something like that) and then if it gets the path-like object, it'll first open the file. Thoughts on that?

That way it would still work in other cases like parsing the mass matrix where we aren't coming directly from a file.

WardBrian replied:
That would be fine by me. Though, note that

> That way it would still work in other cases like parsing the mass matrix where we aren't coming directly from a file.

The metric is never calling this function, it's using csv_bytes_list_to_numpy, which I agree should _not_ accept a file

amas0 replied:
Woops, misread my own code when I made this comment.

Updated the `parse_stan_csv_comments_and_draws` to be able to accept a file path as well and removed the `open` within the stanfit function

WardBrian commented on cmdstanpy/stanfit/mcmc.py:
```diff
@@ -429,83 +430,42 @@ def _assemble_draws(self) -> None:
         self._draws = np.empty(
             (num_draws, self.chains, len(self.column_names)),
-            dtype=float,
+            dtype=np.float32,
```
```suggestion
            dtype=np.float64,
```

amas0 replied:
Everything should be `np.float64` now.

WardBrian commented on cmdstanpy/stanfit/mcmc.py:
```diff
@@ -429,83 +430,42 @@ def _assemble_draws(self) -> None:
+                dtype=np.float32,
+            )
+            for chain in range(self.chains):
+                self._metric[chain] = mass_matrix_per_chain[chain]
```
I think this is equivalent to

```python
if mass_matrix_per_chain[0] is not None:
    self._metric = np.array(mass_matrix_per_chain)
```

amas0 replied:
I was able to clean this bit up. My original parsing code of adaptation lines always returned a 2D array (even in the case of the `diag_e` metric) because that's what the csv parsing code does. I realized this could just be handled in the `parse_hmc_adaptation_lines` code, so now if the `diag_e` metric is used, the parsing will return a 1D array, which allow this to now work:

```py
if all(mm is not None for mm in mass_matrix_per_chain):
    self._metric = np.array(mass_matrix_per_chain)
```

The `all(...)` here will also handle the case when `mass_matrix_per_chain` is an empty list, which happens in the fixed param case.

WardBrian commented on cmdstanpy/stanfit/mcmc.py:
```diff
@@ -429,83 +430,42 @@ def _assemble_draws(self) -> None:
+                    self._step_size[chain],
+                    mass_matrix,
+                ) = stancsv.parse_hmc_adaptation_lines(comments)
+                mass_matrix_per_chain.append(mass_matrix)
```
Thoughts on wrapping these in a try/catch, and then re-raising with the filename if something goes wrong? Otherwise users will just see e.g. `No data found to parse` with no way of knowing which file was bad

amas0 replied:
Done! 

I now raise a `ValueError` with the file name included if any exceptions are thrown in the csv parsing step. It's raised `from exc` with the original exception so the underlying error should propagate through.

WardBrian commented on test/test_stancsv.py:
```diff
@@ -0,0 +1,307 @@
+    with mock.patch.dict("sys.modules", {"polars": None}):
+        with pytest.raises(ValueError):
+            stancsv.csv_bytes_list_to_numpy(lines)
+
```
We have a helper function for this, see e.g.
https://github.com/stan-dev/cmdstanpy/blob/8b37adb96b33a7feac088897060687235f0288f6/test/test_sample.py#L1843-L1861

amas0 replied:
Neat -- moved over to using that.

WardBrian commented on cmdstanpy/utils/stancsv.py:
```diff
@@ -1,17 +1,150 @@
+    if step_size is None:
+        raise ValueError("Unable to parse adapated step size")
+    if matrix_lines:
+        mass_matrix = csv_bytes_list_to_numpy(
```
[comment] already makes me happy to see this is re-using code more than what came before!

WardBrian commented on cmdstanpy/utils/stancsv.py:
```diff
@@ -1,17 +1,150 @@
+    return out
+
+
 def check_sampler_csv(
```
Just noting that, eventually, I'd like to re-write these `check_` functions to consume the same kind of input as the above `parse_` functions (if we don't replace them entirely with a JSON read first, c.f. #685).

Not important for this PR, really just a note to myself.

amas0 replied:
Yeah I think there's a good amount of space to clean up some of these older style functions in light of the new stuff here.

WardBrian reviewed (changes requested):
This is looking really good to me now, thanks!

Don't let the number of comments here be discouraging, most are very small changes. Structurally I think this is really clean now :)

amas0 committed:
Disable pylint warning for re-raising

amas0 committed:
Fixup csv parse typing to 'np.float64'

amas0 committed:
Fixup typing when converting from 'polars.read_csv'

amas0 committed:
Update stancsv tests to np.float64

amas0 committed:
Use 'without_import' helper in 'test_stancsv'

amas0 committed:
Add more testing for non-'diag_e' metric types

amas0 committed:
Clean up mass matrix construction

amas0 committed:
Allow stancsv parse function to accept filename/path

amas0 committed:
Add exception handling to stancsv parsing in assemble_draws

amas0 committed:
Return 1D array when parsing diagnoal hmc mass matrix

amas0 commented:
Okay @WardBrian, I think that's everything addressed? Thanks for the feedback, definitely improved things in a few areas. 

Let me know if there's anything else that stands out from the additions. All test look good (and I added some new ones).

WardBrian commented on cmdstanpy/utils/stancsv.py:
```diff
@@ -1,17 +1,165 @@
+                draws_lines.append(line)
+        return comment_lines, draws_lines
+
+    if isinstance(stan_csv, (str, Path)):
```
[minor] I think it's preferred to use os.PathLike rather than Path for this sort of check

amas0 replied:
Updated

WardBrian commented on cmdstanpy/utils/stancsv.py:
```diff
@@ -1,17 +1,165 @@
+                    has_header=includes_header,
+                )
+                .to_numpy()
+                .astype(np.float64)
```
Rather than doing this afterwards, should we be providing a `schema` (or `schema_override`, can't really tell the difference) argument?

amas0 replied:
I originally didn't the `schema_override` because it requires knowing the number of columns before parsing, but it turns out that you get even more performance out of this is if you do this and disable schema inference entirely. The below is for converting 1000 x 10007 to a numpy array:

```py
%%timeit
n_cols = draws[0].count(b',') + 1
pl.read_csv(io.BytesIO(b"".join(draws)),
            has_header=True,
            schema_overrides=[pl.Float64]*n_cols,
            infer_schema=False
            ).to_numpy()
```
175 ms ± 990 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


```py
%%timeit
pl.read_csv(io.BytesIO(b"".join(draws)), has_header=True).to_numpy().astype(np.float64)
```

225 ms ± 4.48 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

Updated.

WardBrian replied:
My guess is a lot of that speed is less about not trying to infer the schema and more from the fact that polars probably (correctly, I might add) determines that some columns are integers rather than floats, and then the cast actually has to do the conversion. By telling it ahead of time we will be interpreting them as floats always, it avoids the post-hoc recasting

amas0 replied:
Actually if you drop the `infer_schema=False` kwarg, then you lose almost all the benefit:

```py
%%timeit
n_cols = draws[0].count(b',') + 1
pl.read_csv(io.BytesIO(b"".join(draws)),
            has_header=True,
            schema_overrides=[pl.Float64]*n_cols,
            ).to_numpy()
```
220 ms ± 4.65 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

I imagine the casting does have a cost, but at least in my test case where its 10007 columns and only `treedepth__, n_leapfrog__,divergent__` are inferred as ints, it doesn't make up the bulk. It would probably be a different story if a large number of columns happened to be `int`s but I can't imagine that'd be too common?

WardBrian replied:
It depends, people can have a lot of generated quantities that are ints, but certainly not the normal case.

Very interesting!

ahartikainen replied:
is the difference in reading file data or converting to numpy (homogenous vs heterogenous data -> more copying with heterogenous)?

amas0 replied:
The conversion to numpy should be very lightweight for polars, their `to_numpy()` [docs](https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.to_numpy.html) suggest copying data only when necessary. I would guess that with everything inferred as Float64, it would meet the conditions for zero copy. In the heterogeneous case, I wouldn't be surprised if it were only necessary to copy the non-float columns?

In either case, I would guess the main difference is in reading the bytes into the table. My understanding is polars will peek at the first few rows of the csv, attempt to infer the dtype for each column, and then load it in to the table. 

Just kind of speculating though.

WardBrian commented on cmdstanpy/utils/stancsv.py:
```diff
@@ -1,17 +1,165 @@
+    return step_size, mass_matrix
+
+
+def parse_timing_lines(
```
It looks like the only place this new function is called is in the tests?

amas0 replied:
Yeah, this is somewhat vestigial. The parsed timing data was included back when I had the class-based output from the StanCSV. I figured it might be something that could be useful to include as an attribute in the stanfit classes, but it's not something that's used anywhere now. 

Want me to just cut this out and if it's needed in the future, it can be re-implemented?

WardBrian replied:
We currently do read in the output timings elsewhere, so if that code can be replaced with this I'm happy to keep it, otherwise I would remove before merging, yes.

amas0 replied:
The existing timing parsing is pretty deep in some of the `scan_` functions (and not so straightforwardly replaced without digging into those details), so I think it makes sense to just remove this function for now, which I've done.

WardBrian reviewed (approved):
Thanks @amas0!

I have two small questions before merging, but this looks really great

amas0 committed:
Change typing from Path -> os.PathLike

amas0 committed:
Override polars schema inference and set to F64

amas0 committed:
Raise exception if empty list provided to csv_bytes_list_to_numpy

amas0 committed:
Remove unused timing line parsing

WardBrian commented:
Thanks again @amas0!

Looking forward to these utilities spreading to the other methods as well
